ENH add zero_division=nan for classification metrics

marctorsoc opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes https://github.com/scikit-learn/scikit-learn/issues/22625


#### What does this implement/fix? Explain your changes.

This is an extension of https://github.com/scikit-learn/scikit-learn/pull/14900, where I added the parameter `zero_division` for precision, recall, and f1. Afterwards, it was added for jaccard as well.

Here, we add the ability to set zero_division to np.nan, so that np.nan is returned when the metric is undefined. In addition to this:
* when there is an average the numbers that are np.nan (due to undefined and then zero_division) are excluded from the average.
* when beta=0, return precision
* when just one of (precision, recall) is defined and it's 0, return fscore=0. Even if the other metric is undefined.

Specifically:

Precision:

* If pred_sum = 0, undefined
* If average != None, ignore from average any metric being np.nan

Recall:

If true_sum = 0, undefined
If average != None, ignore from average any class metric being np.nan

F-score:

if beta=inf, return recall, and beta=0, return precision
elif precision=0 or recall=0 (or both), return 0.   <------------- *this is a change*
else return zero_division
If average != None, ignore from average any metric being np.nan

Jaccard:

if all labels and pred are 0, return zero_division
If average != None, ignore from average any metric being np.nan


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

marctorsoc committed:
temp commit to checkout sklearn master

marctorsoc committed:
Merge branch 'sklearn_master' into marc_master

# Conflicts:
#	doc/whats_new/v0.21.rst
#	sklearn/metrics/classification.py
#	sklearn/metrics/tests/test_classification.py

marctorsoc committed:
- Changed whats_new to 0.22
- F-score only warns if both prec and rec are ill-defined
- new private method to simplify _prf_divide

marctorsoc committed:
flake8 warnings

marctorsoc committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn

# Conflicts:
#	sklearn/metrics/_classification.py
#	sklearn/metrics/tests/test_classification.py

marctorsoc committed:
first commit. Changes made and tests passing

marctorsoc committed:
- fix linting
- add weights casting to np.array

marctorsoc committed:
add PR number to whats_new

marctorsoc committed:
run black

marctorsoc committed:
tmp commit

marctorsoc committed:
remove all instances of None

marctorsoc committed:
tests fixed

marctorsoc committed:
isort

marctorsoc committed:
Merge branch 'sklearn-main'

marctorsoc committed:
Merge branch 'main' into zero_division_nan

# Conflicts:
#	sklearn/metrics/_classification.py
#	sklearn/metrics/tests/test_classification.py

marctorsoc committed:
merged main into this

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

marctorsoc committed:
remove change to v0.22.rst

marctorsoc committed:
remove change to v0.22.rst

marctorsoc commented on sklearn/metrics/_classification.py:
```diff
@@ -141,6 +145,27 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+def _nan_average(scores: np.ndarray, weights: Optional[np.ndarray]):
+    """
+    Wrapper for np.average, with np.nan values being ignored from the average
+    This is similar to np.nanmean, but allowing to pass weights as in np.average
```
submitted an issue to numpy for this: https://github.com/numpy/numpy/issues/21375, but let me know if there's a better solution than this wrapper!

thomasjpfan replied:
What is the performance difference for running `_nan_average` and `np.average` when there are no nans?

marctorsoc replied:
idk... but I updated this so that if no weights are passed it does just np.nanmean which should be very fast. Otherwise, it creates the mask and does the extra work. tbh, I cannot foresee a big degradation since the operations here are very quick, but 🤷‍♂️

marctorsoc committed:
fix linting and other errors in CI

marctorsoc committed:
fix linting again...

marctorsoc committed:
apply black

marctorsoc committed:
apply black2

marctorsoc committed:
fix docstring examples

marctorsoc commented:
@thomasjpfan and everyone interested, this is now ready to review :)

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -141,6 +147,27 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+
+    scores, weights = scores[~mask], weights[~mask]
+
+    if (weights == 0).all():
+        return np.average(scores)
```
Checking for `weights == 0` adds more computation for an edge case. Can we pass weights directly into `np.average` and not do this check?

marctorsoc replied:
unfortunately
```
ZeroDivisionError
When all weights along axis are zero. 
```
see https://numpy.org/doc/stable/reference/generated/numpy.average.html. But will change into try/except

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -728,10 +755,13 @@ def jaccard_score(
             Calculate metrics for each instance, and find their average (only
             meaningful for multilabel classification).
 
+        Note that if zero_division is np.nan, such values will be excluded
+        from the average.
```
I think we move this into `zero_division`. Currently, when reading the `zero_division` description, the reader needs to scroll up to see what `zero_division=np.nan` does.

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -1243,6 +1303,8 @@ def fbeta_score(
     ``true positive + false negative == 0``, f-score returns 0 and raises
     ``UndefinedMetricWarning``. This behavior can be
     modified with ``zero_division``.
+    Note that if zero_division is np.nan, such values will be excluded
+    from the average.
```
Same here, I think we can place this in the `zero_division` description.

Similar comment for the other docstrings.

thomasjpfan reviewed (commented):
Thank you for the PR!

I think it would be good to see what @jnothman thinks of this `np.nan` behavior.

marctorsoc committed:
PR comments:
- move comment to zero_division
- try/except in nan_average

marctorsoc commented:
> Thank you for the PR!
> 
> I think it would be good to see what @jnothman thinks of this `np.nan` behavior.

@jnothman can I get your 👀  here? and @thomasjpfan if you're happy with current state, maybe approve? :)

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -141,6 +147,28 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
         return sample_score.sum()
 
 
+def _nan_average(scores: np.ndarray, weights: Optional[np.ndarray]):
```
From the code it looks like `weights` can also be a list. I prefer not to have type hints in this file to be consistent.

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -1481,29 +1548,31 @@ def precision_recall_fscore_support(
     Returns
     -------
-    precision : float (if average is not None) or array of float, shape =\
+    precision : float (if `average` is not None) or array of float, shape =\
```
There is already a lot moving parts in this PR. Can this change be a separate small PR?

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -1586,52 +1657,37 @@ def precision_recall_fscore_support(
 
     # warn for f-score only if zero_division is warn, it is in warn_for
-    # and BOTH prec and rec are ill-defined
+    # and BOTH precision and recall are ill-defined
```
This can be reverted so a reviewer can focus on the changes that have to zero division = nan

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -1573,7 +1644,7 @@ def precision_recall_fscore_support(
         true_sum = np.array([true_sum.sum()])
 
-    # Finally, we have all our sufficient statistics. Divide! #
+    # Finally, we have all our sufficient statistics. Divide!
```
This can be reverted so a reviewer can focus on the changes that have to zero division = nan

marctorsoc replied:
Will revert all changes not related to my change. Lmk if you find more

thomasjpfan commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -178,9 +179,9 @@ def test_classification_report_output_dict_empty_input():
-            "f1-score": 0.0,
-            "precision": 0.0,
-            "recall": 0.0,
+            "f1-score": np.nan,
```
This looks like a backward incompatible change for `classification_report`.

marctorsoc replied:
@thomasjpfan well, from my perspective it's a bug fix. Why `macro avg` would return f1-score=np.nan and `weighted avg` 0.0, when there's no input. Should not they return the same?

thomasjpfan replied:
Yea, that makes sense. I am okay with treating this as a bug fix. Can we add another entry on the change log so users know about the behavior change in `classification_report`?

marctorsoc replied:
done

thomasjpfan commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -2537,7 +2564,7 @@ def test_log_loss_pandas_input():
     types = [(MockDataFrame, MockDataFrame)]
     try:
-        from pandas import Series, DataFrame
+        from pandas import DataFrame, Series
```
This can be reverted so a reviewer can focus on the changes that have to zero division = nan

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

marctorsoc committed:
- remove any changes not related to the PR
- move to whats_new v1.2

marctorsoc commented:
Thanks for your comments @thomasjpfan. Applied all of them, except one where I need to be convinced :)

marctorsoc committed:
- update changelog with bug fix

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

marctorsoc commented:
kind reminder here :) @thomasjpfan  @jnothman

thomasjpfan commented on doc/whats_new/v1.2.rst:
```diff
@@ -191,6 +191,19 @@ Changelog
   :pr:`22710` by :user:`Conroy Trinh <trinhcon>` and
   :pr:`23461` by :user:`Meekail Zain <micky774>`.
 
+- |Feature| Added a new possible value for ``zero_division`` to multiple classification
```
```suggestion
- |Feature| Adds  ``zero_division=np.nan`` to multiple classification
```

thomasjpfan commented on doc/whats_new/v1.2.rst:
```diff
@@ -191,6 +191,19 @@ Changelog
+- |Feature| Added a new possible value for ``zero_division`` to multiple classification
+  metrics: :func:`precision_score`, :func:`recall_score`, :func:`f1_score`,
+  :func:`fbeta_score`, :func:`precision_recall_fscore_support`,
+  :func:`classification_report`. This change enables ``zero_division=np.nan``, meaning that when
+  there is a zero division the metric is undefined and should be excluded from averaging.
```
```suggestion
  :func:`classification_report`. With ``zero_division=np.nan`` and
  there is a zero division the metric is undefined and is excluded from averaging.
```

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -141,6 +145,28 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+        return np.nan
+    if weights is None:
+        return np.average(scores[~mask])
+    if isinstance(weights, list):
+        weights = np.array(weights)
```
Nit: This will handle the case for ndarray and list. For ndarray, there is no copy.

```suggestion
    weights = np.asarray(weights)
```

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -141,6 +145,28 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+
+    try:
+        return np.average(scores, weights=weights)
+    except ZeroDivisionError:
+        return np.average(scores)
```
Can we leave a short comment to describe when `ZeroDivisionError` happens?

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -769,12 +798,27 @@ def jaccard_score(
     >>> jaccard_score(y_true[0], y_pred[0])
     0.6666...
 
+    >>> jaccard_score(y_true_with_empty[1], y_pred_with_empty[1])
```
We only use the row indexed 1 in `y_true_with_empty` and `y_pred_with_empty`. I think it will be more obvious to hard code the zero:

```suggestion
    >>> jaccard_score([0, 0, 0], [0, 0, 0])
```

Also I prefer to append this to the end of the Example section. This way, the example flows naturally with the original `y_true` and `y_pred`.

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -731,10 +757,13 @@ def jaccard_score(
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to np.nan, such values will be excluded from the average.
```
Do you think this docstring is consistent with the following behavior?

```python
jaccard_score([0, 0, 0], [0, 0, 0], zero_division=np.nan)
# np.nan
```

In this case, everything is excluded and the return is still nan?

marctorsoc replied:
if everything is excluded there's nothing to compute so the result is undefined, hence np.nan

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

# Conflicts:
#	doc/whats_new/v1.2.rst

marctorsoc committed:
PR comments

thomasjpfan commented on sklearn/metrics/_classification.py:
```diff
@@ -142,6 +146,30 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+        return np.nan
+
+    weights = np.asarray(weights)
+    scores, weights = scores[~mask], weights[~mask]
```
We can early return when there are no nans:

```suggestion
    if not mask.any():
        return np.average(scores, weights=weights)

    scores, weights = scores[~mask], weights[~mask]
```

Here is a quick benchmark:

```python
import numpy as np

n_samples = 100_000
rng = np.random.default_rng(0)
scores = rng.standard_normal(size=n_samples)
weights = rng.poisson(size=n_samples)

%%timeit
_ = _nan_average(scores, weights=weights)
# 400 µs ± 19.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%%timeit
_ = _nan_average_with_early_return(scores, weights=weights)
# 138 µs ± 4.64 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%%timeit
_ = np.average(scores, weights=weights)
# 103 µs ± 4.56 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

There is still a runtime regression when compared with `main`. This type of regression is appears when working with small `n_samples`:

```python
import numpy as np
from sklearn.metrics import precision_recall_fscore_support

rng = np.random.default_rng(42)

n_samples = 100
y_true = rng.integers(low=0, high=100, size=n_samples)
y_pred = rng.integers(low=0, high=100, size=n_samples)
weights = rng.poisson(size=n_samples)

## main
%%timeit
_ = precision_recall_fscore_support(y_true, y_pred, beta=0.5, average="micro")
# 262 µs ± 8.92 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

## PR with above suggestion
%%timeit
_ = precision_recall_fscore_support(y_true, y_pred, beta=0.5, average="micro")
# 300 µs ± 10.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Overall, I think this is acceptable, but I would want to see other opinions.

marctorsoc committed:
np.average when no nans

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

marctorsoc committed:
prevent warning when scores is empty

marctorsoc committed:
fix linting

marctorsoc commented:
@thomasjpfan @betatim @glemaitre @ogrisel can I get someone to review this 🙏 it's been pending for a long time, and I'd like to have it released. I'm happy to work on any changes you request, but I do think it's a useful addition

glemaitre commented:
Could you fix the CIs and move the changelog from 1.2 to 1.3.
I also recall that we recently triage something similar for the MCC: https://github.com/scikit-learn/scikit-learn/issues/25258#issuecomment-1380657067

Could be handy to add it in this PR.

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

# Conflicts:
#	doc/whats_new/v1.2.rst
#	sklearn/metrics/_classification.py

marctorsoc committed:
move changelog changes to 1.3

marctorsoc committed:
fix tests

marctorsoc committed:
restore second entry for the bug fix

marctorsoc committed:
add support for MCC

marctorsoc commented:
> Could you fix the CIs and move the changelog from 1.2 to 1.3. I also recall that we recently triage something similar for the MCC: [#25258 (comment)](https://github.com/scikit-learn/scikit-learn/issues/25258#issuecomment-1380657067)
> 
> Could be handy to add it in this PR.

@glemaitre  done: added support for MCC + fixed CI

I added another entry to the changelog, but happy to write it all in one or two. Just lmk

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -48,11 +48,15 @@
-        return
+        return np.float64(zero_division)
+    elif np.isnan(zero_division):
+        return np.float64(np.nan)
```
I think `np.nan` is already a `np.float64`. I don't think that we need to pass it in the constructor.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -48,11 +48,15 @@
+        return np.float64(np.nan)
     raise ValueError(
-        'Got zero_division={0}. Must be one of ["warn", 0, 1]'.format(zero_division)
+        'Got zero_division={0}. Must be one of ["warn", 0, 1, np.nan]'.format(
+            zero_division
```
Since we modify this one, I think that we can use an f-string.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -143,6 +147,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+    """
+    Wrapper to combine np.average and np.nanmean, so np.nan values are ignored
+    from the average and weights can be passed. Note that when possible,
+    we degenerate to the prime methods.
```
I assume `delegate` is what you mean by `degenerate`

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -143,6 +147,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+    if weights is None:
+        return np.nanmean(scores)
+
+    weights = np.asarray(weights)
```
Let's make it explicit that we don't need a copy if this is already an array.

```suggestion
    weights = np.array(weights, copy=False)
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -143,6 +147,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
+    if mask.all():
+        return np.nan
+
+    if weights is None:
+        return np.nanmean(scores)
```
A bit link with the comment of @thomasjpfan, I am wondering if we should precompute `mask.any()` and use the basic mean when no `nan` are present. It should avoid some overhead:

```python
is_any_nam = mask.any()
if weights is None:
    return np.nanmean(scores) if is_any_nan else np.mean(scores)
elif not is_any_nan:
    return np.average(scores, weights=weights)
```

marctorsoc replied:
this does not work well because when scores sums 0, np.average throws a `ZeroDivisionError`

glemaitre replied:
Oh yes, right.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -763,10 +793,13 @@ def jaccard_score(
         Sample weights.
 
-    zero_division : "warn", {0.0, 1.0}, default="warn"
+    zero_division : "warn", 0.0, 1.0, np.nan, default="warn"
```
Here and after:

```suggestion
zero_division : "warn", {0.0, 1.0, np.nan}, default="warn"
```

We can consider that this is a set of the possible value to pass.

marctorsoc replied:
I will assume you meant `zero_division : {"warn", 0.0, 1.0, np.nan}, default="warn"`

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -763,10 +793,13 @@ def jaccard_score(
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to np.nan, such values will be excluded from the average.
```
```suggestion
        - If set to `np.nan`, such values will be excluded from the average.
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -786,7 +819,7 @@ def jaccard_score(
     positives for some samples or classes. Jaccard is undefined if there are
     no true or predicted labels, and our implementation will return a score
-    of 0 with a warning.
+    of 0 with a warning. This behaviour can be modified with `zero_division`.
```
```suggestion
    of 0 with a warning. This behaviour can be modified by setting the parameter
    `zero_division`.
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -823,10 +864,19 @@ def jaccard_score(
     >>> jaccard_score(y_true, y_pred, average=None)
     array([1. , 0. , 0.33...])
+
+    Finally, see how zero_division works:
```
```suggestion
    Finally, you can use `zero_division` to set the output when the metric is undefined:
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -896,6 +946,13 @@ def matthews_corrcoef(y_true, y_pred, *, sample_weight=None):
     sample_weight : array-like of shape (n_samples,), default=None
         Sample weights.
 
+    zero_division : "warn", 0, 1 or np.nan, default="warn"
```
same remark as above

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -929,6 +986,13 @@ def matthews_corrcoef(y_true, y_pred, *, sample_weight=None):
+    >>> import numpy as np
+    >>> y_true = [1, 1, 1, 1]
+    >>> y_pred = [1, 1, 1, 1]
+    >>> matthews_corrcoef(y_true, y_pred, zero_division=1.0)
```
It could be worth to have the default

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1163,10 +1239,17 @@ def f1_score(
+    >>> y_pred_empty = [0, 0, 0, 0, 0, 0]
+    >>> f1_score(y_true_empty, y_pred_empty)
+    0.0...
+    >>> f1_score(y_true_empty, y_pred_empty, zero_division=1)
```
```suggestion
    >>> f1_score(y_true_empty, y_pred_empty, zero_division=1.0)
```

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -724,9 +725,30 @@ def test_cohen_kappa():
+    "zero_division",
+    ["warn", 0, 1, np.nan],
+)
+def test_matthews_corrcoef_nan(zero_division):
```
Could we make a common test for all metrics that support this feature.

I would write two parameterized tests: (i) specifically for warn and (ii) another for the other.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -2395,7 +2490,7 @@ class 2       1.00      0.67      0.80         3
         report_dict = {label[0]: label[1:] for label in rows}
         for label, scores in report_dict.items():
-            report_dict[label] = dict(zip(headers, [i.item() for i in scores]))
+            report_dict[label] = dict(zip(headers, map(float, scores)))
```
can we use the list comprehension instead of the `map` function

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -2427,7 +2522,7 @@ class 2       1.00      0.67      0.80         3
 
         if output_dict:
-            report_dict[line_heading] = dict(zip(headers, [i.item() for i in avg]))
+            report_dict[line_heading] = dict(zip(headers, map(float, avg)))
```
same here.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1597,9 +1691,11 @@ def precision_recall_fscore_support(
+    zero_division_value = _check_zero_division(zero_division)
+
+    if beta < 0 or np.isnan(beta):
+        raise ValueError("beta should be >=0 (np.inf is ok) in the F-beta score")
```
```suggestion
        raise ValueError(
            f"beta should be >=0 in the F-beta score. Got {beta!r} instead."
        )

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -950,7 +1014,14 @@ def matthews_corrcoef(y_true, y_pred, *, sample_weight=None):
+                "control this behaviour."
+            )
+            warnings.warn(msg, UndefinedMetricWarning, stacklevel=2)
+            return 0.0
```
We should also be returning `zero_division` that should set to `0.0`

glemaitre replied:
I mean we should be calling `_check_zero_division` to get the validated value to return.

glemaitre reviewed (commented):
A first round of review.

marctorsoc committed:
tmp commit

marctorsoc committed:
PR comments, round 1

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

marctorsoc commented:
@glemaitre I went over all your comments and either applied or answered

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

# Conflicts:
#	doc/whats_new/v1.3.rst

marctorsoc committed:
fix linting

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -724,9 +725,39 @@ def test_cohen_kappa():
-def test_matthews_corrcoef_nan():
-    assert matthews_corrcoef([0], [1]) == 0.0
-    assert matthews_corrcoef([0, 0], [0, 1]) == 0.0
+@pytest.mark.parametrize("zero_division", [0, 1, np.nan])
+@pytest.mark.parametrize(
+    "y_true, y_pred",
+    [
+        ([0], [1]),
+        ([0, 0], [1, 1]),
+        ([], []),
+    ],
+)
+def test_matthews_corrcoef_nan(zero_division, y_true, y_pred):
+    with warnings.catch_warnings(record=True) as record:
+        mcc = matthews_corrcoef(y_true, y_pred, zero_division=zero_division)
+        assert not record
+
+    if np.isnan(zero_division):
+        assert np.isnan(mcc)
+    else:
+        assert mcc == zero_division
+
+
+@pytest.mark.parametrize(
+    "y_true, y_pred",
+    [
+        ([0], [1]),
+        ([0, 0], [1, 1]),
+        ([], []),
+    ],
+)
+def test_matthews_corrcoef_nan_warn(y_true, y_pred):
+    with warnings.catch_warnings(record=True) as record:
+        mcc = matthews_corrcoef(y_true, y_pred, zero_division="warn")
+        assert len(record) == 1
+        assert mcc == 0.0
```
What I meant is the following general test:

```python
@pytest.mark.parametrize("zero_division", [0, 1, np.nan])
@pytest.mark.parametrize("y_true, y_pred", [([0], [0]), ([], [])])
@pytest.mark.parametrize("metric", [
    jaccard_score,
    matthews_corrcoef,
    f1_score,
    partial(fbeta_score, beta=1),
    precision_score,
    recall_score,
])
def test_zero_division_nan_no_warning(metric, y_true, y_pred, zero_division):
    """Check the behaviour of `zero_division` when setting to 0, 1 or np.nan.
    No warnings should be raised.
    """
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        result = metric(y_true, y_pred, zero_division=zero_division)

    if np.isnan(zero_division):
        assert np.isnan(result)
    else:
        assert result == zero_division


@pytest.mark.parametrize("y_true, y_pred", [([0], [0]), ([], [])])
@pytest.mark.parametrize("metric", [
    jaccard_score,
    matthews_corrcoef,
    f1_score,
    partial(fbeta_score, beta=1),
    precision_score,
    recall_score,
])
def test_zero_division_nan_warning(metric, y_true, y_pred):
    """Check the behaviour of `zero_division` when setting to "warn".
    A `UndefinedMetricWarning` should be raised.
    """
    with pytest.warns(UndefinedMetricWarning):
        result = metric(y_true, y_pred, zero_division="warn")
    assert result == 0.0
```

Here, I don't include `precision_recall_f1_score_support` and `classification_report` because they are a bit different.

The idea is not to test all potential cases that trigger the warning but the general behaviour when the `UndefinedMetricWarning` is triggered.

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1678,36 +1712,53 @@ def test_precision_recall_f1_score_multilabel_2():
 
 @ignore_warnings
-@pytest.mark.parametrize("zero_division", ["warn", 0, 1])
+@pytest.mark.parametrize("zero_division", ["warn", 0, 1, np.nan])
```
I think that we can directly set the behaviour outside without any programming.

```suggestion
@pytest.mark.parametrize(
    "zero_division, zero_division_expected",
    [("warn", 0), (0, 0), (1, 1), (np.nan, np.nan)]
)
```

glemaitre commented on doc/whats_new/v1.3.rst:
```diff
@@ -118,6 +118,42 @@ Changelog
+- |Fix| :class:`feature_extraction.FeatureHasher` raises an informative error
+  when the input is a list of strings. :pr:`25094` by `Thomas Fan`_.
+
+:mod:`sklearn.metrics`
+.................................
```
```suggestion
:mod:`sklearn.metrics`
......................
```

glemaitre commented:
So I look a bit at the other classification metrics and I have the following question:

- `accuracy_score` and `balanced_accuracy` will raise a warning when `y_true` and `y_pred` are empty. We are not really consistent with other metrics then.
- `cohen_kappa_score` seems to be a good candidate that should be added to the list of metrics to accept a `zero_division` parameter.
- `class_likelihood_ratio` has a `raise_warning` parameter that is inconsistent with the current PR. It should not be addressed in this PR. But we should investigate if the deprecation cost is lower than making this metric consistent (and check that it makes sense indeed).

@thomasjpfan could you provide your input on those three different questions?
Otherwise the PR is on a good state on my side.

marctorsoc committed:
PR comments + kappa score

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

marctorsoc commented:
> So I look a bit at the other classification metrics and I have the following question:
> 
> * `accuracy_score` and `balanced_accuracy` will raise a warning when `y_true` and `y_pred` are empty. We are not really consistent with other metrics then.
> * `cohen_kappa_score` seems to be a good candidate that should be added to the list of metrics to accept a `zero_division` parameter.
> * `class_likelihood_ratio` has a `raise_warning` parameter that is inconsistent with the current PR. It should not be addressed in this PR. But we should investigate if the deprecation cost is lower than making this metric consistent (and check that it makes sense indeed).
> 
> @thomasjpfan could you provide your input on those three different questions? Otherwise the PR is on a good state on my side.

Added (2). (1) and (3) seemed more convoluted. Lmk your thoughts

marctorsoc commented:
@glemaitre @thomasjpfan kind reminder about this :)

glemaitre commented:
The CI is broken. @marctorsoc Could you solve that.

marctorsoc committed:
Merge branch 'sklearn-main' into zero_division_nan

# Conflicts:
#	doc/whats_new/v1.3.rst

marctorsoc committed:
fix sklearn/tests/test_public_functions.py

marctorsoc committed:
format with black

marctorsoc committed:
again

marctorsoc committed:
remove Hidden

marctorsoc commented:
> The CI is broken. @marctorsoc Could you solve that.

@glemaitre done

thomasjpfan commented:
@glemaitre I'll say to standardize over `zero_division` as much as possible, which includes all the metrics you suggested in https://github.com/scikit-learn/scikit-learn/pull/23183#issuecomment-1400159535.

As for this PR, I think it covers a lot, which makes it harder to reviewer. If we can break up this PR into smaller PRs that adds `zero_division` one metric at a time, I think it will speed up the process.

marctorsoc commented:
> @glemaitre I'll say to standardize over `zero_division` as much as possible, which includes all the metrics you suggested in [#23183 (comment)](https://github.com/scikit-learn/scikit-learn/pull/23183#issuecomment-1400159535). I think it's worth deprecating `raise_warning` in `class_likelihood_ratio`.
> 
> As for this PR, I think it covers a lot, which makes it harder to reviewer. If we can break up this PR into smaller PRs that adds `zero_division` one metric at a time, I think it will speed up the process.

@thomasjpfan I don't think one metric per PR is a) feasible, b) worth it.  

a) because e.g. precision and recall both affect classification_report. b) because it would mean 10+ PRs with very few changes each.

I propose the following:
1. PR to add zero_division=np.nan to precision, recall, f1, fbeta_score, precision_recall_fscore_support and classification_report
2. PR to add zero_division to `matthews_corrcoef` and :func:`cohen_kappa_score`.
3. PR to fix :func:`classification_report` so that empty input will return `np.nan` 
4. PR to add zero_division for `accuracy_score` and `balanced_accuracy` which raise a warning when y_true and y_pred are empty
5. PR to add zero_division for class_likelihood_ratios and remove `raise_warning`

I can take care of of most if not all of these, but first I want a greenlight as this will take me some time

glemaitre commented:
I can commit time reviewing individual PRs as proposed by @marctorsoc True that it is difficult to dissociate all metrics.

marctorsoc commented:
@glemaitre @thomasjpfan please see https://github.com/scikit-learn/scikit-learn/pull/25531

erikhuck commented:
@glemaitre are we able to get this merged at some point or is it being split into different pull requests?

glemaitre commented:
It is split in different pull-requests.

marctorsoc commented:
I'll close this PR to avoid confusion
